### ISSUE TITLE
additional environment variables

### ISSUE THREAD

**Issue by terencode** [AladW/aurutils#302]
This (JSON?) configuration file could first hold per-repository-settings:

- You could add different settings per repo:
    -> `devel` to know it is holding vcs packages
    -> `default` for the default repo to use when for example adding a new package

**Comment by He-Ro** [comment]
Just leaving my two cents: A configuration seems like a reasonable feature request, but I would suggest not to use JSON format. One simple reason is that there is no way to add comments to a JSON file, which is IMO quite an important feature in a configuration file. One alternative might be [TOML](https://github.com/toml-lang/toml), which looks a lot like the INI format, but is somewhat standardized.

**Comment by terencode** [comment]
@He-Ro ok, did not know that.

**Comment by AladW** [comment]
Instead of implementing a TOML parser, why not use environment variables? Similar to what's already done with `AURDEST`.

**Comment by terencode** [comment]
@AladW I was thinking of a configuration file if at some point there would really be the need of one, but I guess using environment variables is just fine for now.

**Comment by terencode** [comment]
When #283 will be merged, I will face a situation where I have some devel packages getting updated daily, and others only once every few weeks. Maybe it would be interesting to have per-package settings for in this case ignore some devel packages or even better, set up a database holding last time a devel package has been checked, allowing to configure per-package checking frequency.

**Comment by AladW** [comment]
From IRC:

>13:31 <Terence> alad: could you please check my last message on #302 ? I might have went a bit too far but I thought it was worth thinking about it.
13:32 <alad> databases are complicated :P
13:32 <alad> but if you already know which packages you want to update frequently, you can just specify those to aursync instead of using -u
13:33 <alad> that should work equally well with the proposed --devel flag
13:33 <Terence> "databases are complicated :P" > only when you are dealing with 10000+ entries and have to setup relational stuffs :P
13:34 <Terence> " but if you already know which packages you want to update frequently, you can just specify those to aursync instead of using -u" what if it's the reversed thing, you want to updae everything frequently BUT a few packages ?
13:35 <Terence> I was using pacaur before, and it was a pain for me when I had this situation
13:36 <alad> well aursync -u just adds whatever aurvercmp gives you. so you can just run aurvercmp and filter out the things you don't want
13:36 <alad> I should add some more example pipelines detailing this
13:36 <Terence> oh that is really cool
13:37 <Terence> would definitly solve this problem
13:37 <alad> that's the benefit over monolothic helpers like pacaur, you can easily adjust any component of the workflow

**Comment by AladW** [comment]
One useful environment variable may be `$AUR_LIB_DIR`, to easily override shipped scripts. (compare git's `--exec-dir`). Shouldn't be problematic with sudoers rules since the env is discarded unless `SETENV` is available.

**Comment by terencode** [comment]
@AladW Makes sense to me, what about AUR_REPO_DEVEL ?

**Comment by AladW** [comment]
Flagging a repository as containing development packages (through `AUR_REPO_DEVEL`) or otherwise sounds better than implementing some arbitrary pattern like "(git|hg|bzr|svn)'. 

You can then just dump all vcs packages into a specific repo like `custom-vcs`. Thanks to the new `aur sync --provides`, this repo could _only_ contain such packages.

**Comment by maximbaz** [comment]
Having a repo for VCS packages is an interesting idea, so far I didn't use it because of lack of `--provides` and because it's annoying to specify repo every time, if you have more than one repo.

How do you see `AUR_REPO_DEVEL` variable be used? When I sync a new package, will `aur sync` be guessing whether this package belongs to a default repo or this devel repo, and how would it do so? How would `AUR_REPO_DEVEL` affect how `aur sync -u` works?

**Comment by AladW** [comment]
The workflow would be that the user decides whether a package is a development package, e.g.
```
export AUR_REPO=custom
aur sync something-release
aur sync --repo custom-vcs --devel something-git
```
The idea was that `AUR_REPO_DEVEL` is something that's checked after a repository has been selected. When trying to answer your question I noticed some limitation there:
```
repo=${repo:-$AUR_REPO}
# how to check from here that repo is devel?
```
I guess you could use an (associative) array that contains all "devel" repos, but that's limited to bash/zsh and not everyone uses those as interactive shell. 
```
repo=${repo:-$AUR_REPO}
if [[ ${AUR_REPO_DEVEL[$repo]} ]]; then
   # do devel stuff
fi
```
For a better UI you could have `--devel` take an optional argument instead, being the VCS repo. That relates "stuff done with `--devel`" with "the repo containing devel packages" and doesn't require you to specify two flags for one thing.
```
export AUR_REPO=custom
aur sync something-release
aur sync --devel custom-vcs something-git
```
That leaves the question, "what if you want a single repo for everything?"
----
An alternative is specifying all these things in some configuration file like `repoctl` does. I'd like to see how far we can get with the above approach before considering that though.

**Comment by maximbaz** [comment]
> The idea was that `AUR_REPO_DEVEL` is something that's checked after a repository has been selected. 

But I thought the whole point of defining `AUR_REPO` and `AUR_REPO_DEVEL` is to avoid manually providing repos ever again.

I see it in the following way:

```bash
export AUR_REPO=custom
export AUR_REPO_DEVEL=custom-devel

aur sync something-release      # no --devel flag provided, put the packages in AUR_REPO
aur sync --devel something-git  # --devel flag provided, put the packages in AUR_REPO_DEVEL
```

And then I imagine the update runs in the following way:

```bash
aur sync -u           # no --devel flag provided, don't use aur-srcver, just update packages that exist in AUR_REPO
aur sync --devel -u   # --devel flag provided, list packages in AUR_REPO_DEVEL, run aur-srcver on them, then update the listed packages
```

That way you don't need to configure PCRE patterns, whatever was first sync'ed with `--devel` is considered a devel package.

However, I'm only not sure about dependencies, for example if I do:

```bash
aur sync --devel something-git
```

And in addition to `something-git` I am prompted to install `dependency1-release` and `dependency2-git`, into which repo will they be added?

**Comment by terencode** [comment]
> And in addition to something-git I am prompted to install dependency1-release and dependency2-git, into which repo will they be added?

That's the problem I am facing right now: I want to install tilix-git to my aur-git repo and it wants to install gtkd to my git repo instead of gtkd-git.

**Comment by AladW** [comment]
If the -git package is already in your devel repo, `--provides` should suffice. If it's not (or in the case of gtkd, the release package is in the official repos), the -git deps should either be installed first or we'd need some way to find out if said deps have -git versions available. 

Latter requires AurJson to support providers. There seems to be some progress on this, last upstream response:

https://lists.archlinux.org/pipermail/aur-dev/2018-March/004495.html

I don't think adding different packages to different repos in the same step is a great idea, since that implies repos are no longer self-sufficient.

**Comment by terencode** [comment]
@AladW Ok, this only happened when adding it for the first time.

**Comment by maximbaz** [comment]
> I don't think adding different packages to different repos in the same step is a great idea, since that implies repos are no longer self-sufficient.

But isn't `--provides` made for using non-self-sufficient repositories anyway? 🙂

It's just strange to see a repo that was meant to hold only devel packages, but which actually holds devel packages as well as some of their dependencies, even non-devel ones. Over time things like this could happen:

1. Install `aur sync --devel pkg-git`, it puts `pkg-git` and its dependency `dep-release` in `AUR_REPO_DEVEL`.
1. Install `aur sync --provides pkg2-release` (also depends on `dep-release`), which puts `pkg2-release` in `AUR_REPO`, but doesn't put `dep-release` in there because it already exists in `AUR_REPO_DEVEL` (is that how `--provides` work?).
1. Remove `pkg-git` from the repo, `dep-release` stays in `AUR_REPO_DEVEL` even though it is not a devel package and it is only needed by `pkg2-release` which is also a non-devel package.

**Comment by AladW** [comment]
|But isn't --provides made for using non-self-sufficient repositories anyway? 

Well, no, since it only queries the repo you're building to. (see `aur rfilter -d`). It doesn't go across repo boundaries.

**Comment by maximbaz** [comment]
Could you explain this then?

> You can then just dump all VCS packages into a second repo like custom-vcs. Thanks to the new aur sync --provides, this repo could _only_ contain such packages.

How, since `--provides` is buind by the repo, wouldn't it also contain all dependencies, even non-devel ones?

-----------

So based on how I currently understand the workings of aurutils, I see some downsides of having a separate repo for devel packages:
- It contains not only devel packages, but also all their dependencies, so it's not really correct to say "package in `AUR_REPO_DEVEL` - then it is devel". We can't use that for example to decide whether to run `aur-srcver` or not.
- Dependencies are duplicated between repos, i.e. if a dependency is needed by packages in both `AUR_REPO` and `AUR_REPO_DEVEL`, the dependency will be built twice and stored twice in each repo.

What are the actual benefits of having multiple repos?

**Comment by AladW** [comment]
Well, let's assume the AUR contains both the development and the release version. (otherwise you just install the deps from the arch repos unless you specifically want to build a devel version yourself) Since the AUR has no ability to search by provides (at least not outside the web interface), you'd build the -git dependencies first, add to the vcs repo, and then continue with the usual process.

>What are the actual benefits of having multiple repos?

The idea was to lose the semantics on what a devel package is by simply moving those that may be considered as such (by the user) to a differernt location. Though I can see now that `AUR_REPO_DEVEL` itself is semantics.

The question is, why all this effort in the first place. Probably because aur-srcver being such a slow program. :disappointed:

**Comment by maximbaz** [comment]
> Probably because aur-srcver being such a slow program.

Basically you say, if we make `aur-srcver` fast enough, we can treat all packages as devel and problem solved, right? Then let's see if `aur-out-of-date` helps with the performance, and maybe that's all that is needed?

**Comment by AladW** [comment]
Pretty much. Note that the existing `--ignore` flag could also be leveraged, since it removes candidate packages early on. I guess you might as well extend it to take a pattern.

**Comment by maximbaz** [comment]
I've realized that if we end up using `aur-out-of-date`, we would not be able to treat all packages as devel, because the purpose of `aur-out-of-date` is to indicate packages that require _maintainers_ to update their PKGBUILDs when a new release of an app is published. 

If we start using `aur-out-of-date` on all packages, we will see "update available" even before the PKGBUILD on AUR is updated for this new version. So `aur-out-of-date` is only useful for devel packages, and we still have to distinguish the two types... 

Here's the hardcoded definition of [what an VCS package is](https://github.com/simon04/aur-out-of-date/blob/08faebe65262e4bd0a0dff6a9f529585094bd890/main.go#L98) in `aur-out-of-date`:

```golang
isVcsPackage := strings.HasSuffix(pkg.Name(), "-git") || strings.HasSuffix(pkg.Name(), "-hg") || strings.HasSuffix(pkg.Name(), "-svn")
```

**Comment by maximbaz** [comment]
> > Basically you say, if we make aur-srcver fast enough, we can treat all packages as devel and problem solved, right? Then let's see if aur-out-of-date helps with the performance, and maybe that's all that is needed?
>
> Pretty much. Note that the existing --ignore flag could also be leveraged, since it removes candidate packages early on. I guess you might as well extend it to take a pattern.

Since perf on `aur-srcver` is improved as much as we can, I did some measurements.

`aur-srcver ~/.cache/aurutils/sync/*-git`:
- Number of packages: 9
- Run time: 2 sec

`aur-srcver ~/.cache/aurutils/sync/*`

- Number of packages: 83
- First run took 2 min 33 sec.
- Second run took 1 min 53 sec (huge projects like `chromium-vaapi` take particularly long time to finish)
- After the first run, the size of my cache folder grew from 3.1GB to 11.3GB

My summary: 

1. No, we cannot treat all packages as devel, querying RPC is still much much more efficient.
2. I don't see much sense in having two repos for regular and VCS packages, because many VCS packages depend on regular AUR packages, which means a lot of packages will be built twice, for both repos.

I've been using #283 as my daily driver for a couple of months now, I think that's the best way to proceed. I have one repo, I can use `aur sync -u` to update only regular AUR packages and `aur sync -u --devel` to update both regular and VCS packages, however with the latest perf improvements the difference is negligible, so I always use `--devel`.

Speaking of new environment variables, we could create one to allow overriding the definition of VCS packages, and hardcode the default value of `-(bzr|git|hg|svn)$`.

**Comment by terencode** [comment]
@maximbaz So the idea is you have one repo for both VCS and regular packages, then you apply a filter  (that you can customize) when you using --devel on the package names to differentiate the devel ones from regulars ones ?

**Comment by maximbaz** [comment]
Correct. 

One repo, because regular and VCS packages often intersect and have dependencies on each other, with one repo you don't have build them twice (if you are using `--provides`, I use it always). 

When you are not using `--devel`, only RPC is being queried, and thus VCS packages are never getting into the list of packages that need an update, because RPC doesn't know their true latest version. 

With `--devel` in #283, in addition to querying RPC, the `aur-srcver` script is also being executed on a subset of packages. The `aur-srcver` is slower, but is able to retrieve the true latest versions. Currently in #283 `--devel` only executes `aur-srcver` on `-git$` packages, because I only know of two people who are that branch on daily basis (one of them is me), and both of us only have `-git` packages anyway 😉 But the idea is to make this configurable and have a better default (like `-(bzr|git|hg|svn)$`), I'm just waiting for a suggestion on how this should be done best - configurable via an environment variable, configurable via an extra flag like `--vcs "-(bzr|git|hg|svn)$"`, or how.

**Comment by terencode** [comment]
Ok, I would think using environment variable is better as it's not something you want to specify each time.

**Comment by maximbaz** [comment]
I think so too, I pushed a commit that uses `-(bzr|git|hg|svn)$` as a new default value and allows configuring this via a new environment variable which for now I called `AURVCS`.

**Comment by terencode** [comment]
What about being able to make '- -continue', '--no-reset' '--no-view' etc. the default so they are specified automatically when you use aur sync? Maybe in a similar way youtube-dl does it? https://github.com/rg3/youtube-dl#configuration

**Comment by maximbaz** [comment]
Personally [I just went for aliases](https://github.com/maximbaz/dotfiles/blob/739e5e46e950291fb08ad651d16b7ab29f6764f1/.zsh/pacman.zsh#L12-L14), as of now there are really only 3 commands that I need on regular basis: sync a new package, update my packages and build a PKGBUILD in the current directory.

```bash
alias aurs='aur sync -sc --provides'
alias auru='aur sync -scu --devel --provides'
alias aurb='aur build -sc -d maximbaz-aur'
```

I don't know for sure, but I would think most people do something like that 🙂

**Comment by terencode** [comment]
@maximbaz Ok, I think this makes sense as long as there are not that many options.

**Comment by AladW** [comment]
At some point I might add an equivalent to `yaourt --sucre` (`aur sync --bloat`?) that combines some common options that shouldn't be enabled by default.

**Comment by AladW** [comment]
`AUR_REPO` (aur-build, aur-sync) and `AUR_PAGER` (aur-sync) are now implemented, the rest should be discussed in #283.
